p2p/simulations/network.go

```diff
@@ -718,17 +721,97 @@ func (net *Network) Load(snap *Snapshot) error {
 			return err
 		}
 	}
+
+	// Prepare connection events counter.
+	allConnected := make(chan struct{})     // closed when all connections are established
+	eventLoopStarted := make(chan struct{}) // ensures that event loop is started before it is closed
+	done := make(chan struct{})             // ensures that the event loop goroutine is terminated
+	defer close(done)
+
+	go func() {
+		// Subscribe to event channel.
+		events := make(chan *Event)
+		sub := net.Events().Subscribe(events)
+		defer sub.Unsubscribe()
+
+		// Expected number of connections.
+		total := len(snap.Conns)
+		// counter tracks the current number of connections.
+		var counter int
+
+		// once is a closed channel that is read in the event loop below
+		// only once.
+		// It ensures that eventLoopStarted is closed which signals that
+		// it is safe to call connect method on the network without the
+		// possibility to miss a few first connection events.
+		once := make(chan struct{})
+		// Close once channel so that it can be read from in the event loop.
+		close(once)
+
+		for {
+			select {
+			case e := <-events:
+				// Detect only connection events.
+				if e.Type != EventTypeConn {
+					continue
+				}
+				// Detect only "connect" events of all connection events.
+				if !e.Conn.Up {
+					continue
+				}
+				// Check that the connection is from the snapshot.
+				for _, conn := range snap.Conns {
```

Comment:
I suppose that you meant by the third event: _B connects to C_?

I assumed that disconnections won't happen during load, since the snapshot is already created from a stable network. If all connections need to be verified, that can be done on every connection event instead of heaving a counter, or keeping a record of expected connections and track their _up_ state.

Reply:
Thanks Viktor.

Reply:
@zelig why should this ever happen while we are booting a snapshot?